cms_coding_rules.md

```diff
@@ -0,0 +1,144 @@
+
+
+## Outline
+#### 1 -- Introduction
+#### 2 -- Naming Rules
+#### 3 -- Style Rules
+#### 4 -- Technical Coding Rules
+#### 5 -- Documentation Rules
+#### 6 -- Packaging Rules
+#### 7 -- Design and Coding Guidelines
+
+## 1 -- Introduction
+This document describes the CMS C++ software naming, coding, style, and documentation rules and recommendations.
+All CMS C++ software is expected to comply with the rules. The asterisk (\*) after some rules indicates that there may be exceptional use cases where the rule may be violated with good justiﬁcation.
+Coding rules are meant to prevent serious problems in software function, performance, maintainability, usability, and portability. The Packaging Rules section also has some brief guidelines for Python configuration.
+
+## 2 -- Naming Rules
+1. C++ header ﬁles use the sufﬁx .h, e.g. `CaloCluster.h`. (*)
+2. For C++ source ﬁles, the preferred sufﬁx  is .cc, e.g. `CaloCluster.cc`, though there are existing files using .cpp and .cxx suffixes. (*)
+3. For a header file that contains a class, name that ﬁle after the class.
+4. Name source ﬁles after the class.
+5. For class, struct, type, and enumeration names use upper class initials, e.g. `GeometryBuilder`.
+6. For namespaces use lower case, e.g. `namespace edm`.
+7. Start method names with lowercase, use upper case initials for following words, e.g. `collisionPoint()`.  
+  Allowed exception: Implementation of virtual methods inherited from external packages, e.g. `ProcessHits()` method required by Geant4.  
+8. Start data member names with lower case. A trailing "\_" is the preferred method to distinguish a data member from the getter method (e.g. `momentum_`).
+9. Using "set" for a setter method is preferred, e.g. `setMomentum(double m)`.
+10. For a getter method, using the value name is preferred, e.g. `momentum()`.
+11. Do not use single character names, except for loop indices.
+12. Do not use special characters, except for "_" where allowed.
+13. Do not use "_" as ﬁrst character. Only use it as the last character for class data member names, not local variable names.
+14. Do not use "__".
+15. Use clear and explanatory variable names.
+
+## 3 -- Style Rules
+1. Do not indent pre-processor directives -- there should be no leading spaces before a directive.  (*)
+2. Never change the language syntax using `#deﬁne`.
+3. Do not use spaces between a function or method name and its argument list, e.g. `foo()` rather than `foo ()`.
+4. Do not use spaces in front of [] or on either side of -> . For example, `vector[i]` instead of `vector [i]`.
+5. Separate expressions in a `for` statement by spaces.
+6. Use the same indentation for comments as for the block the comments refer to.
+
+## 4 -- Technical Coding Rules
+1. Protect each header ﬁle from multiple inclusion with:  
+`#ifndef PackageName_SubPackageName_FileName_h  
+#define PackageName_SubPackageName_FileName_h`  
```

Comment:
Should use three backticks to show the newline
````
```
#ifndef PackageName_SubPackageName_FileName_h  
#define PackageName_SubPackageName_FileName_h
```
````

Reply:
I split the line but could not get the three back-quotes to work. I had to back-quote each line separately. I think it looks OK now.